### PR TITLE
feat(project): run Neo4j migrations before starting the server

### LINKED ISSUE #113: Add support to run cypher queries on CE startup
In the current readme: https://github.com/trompamusic/ce-api/blob/5a389e07bb8fc5828e7bc2e3783a3fce6e3b8d5e/README.md#search-indexes

> TODO: this should be added to the startup process eventually.

We also have a requirement to run cypher queries during startup for the Annotation schema: https://github.com/trompamusic/ce-api/pull/100/files#r391106249

so we would like to be able to have a hook where we can write some queries to have these commands run each time on startup to ensure that the relevant items are created if necessary

### PR DESCRIPTION
**Pull request by ChristiaanScheermeijer**
@alastair since we needed to implement the `createNodeIndex` migrations anyways, I've added a generic migrations runner before starting the server.

A migration file should export an `async` function. This function will receive the `transaction` session and the driver instance in the arguments.

Use the `transaction` session for changes that will be made to the schema or data so it will be reverted when a migration fails.

If you need to verify data first, you can use the `driver.session()` session in order to read from the DB before running transaction queries.

```js
export default async function(transaction, driver) {
  await driver.session().run('MATCH (n:Person) RETURN n')

  // will be queued
  transaction.run('CREATE ....')
}
```

Fixes #113

### CODE REVIEW

**Comment by alastair** [comment]
When running this on a fresh database I see

```
api_1    | Running migrations
api_1    | ------------------
api_1    | Verifying Neo4j connectivity...
api_1    | Connected to Neo4j instance, running migrations...
api_1    | 1. /app/src/migrations/01-drop-search-indices.cypher
api_1    | 2. /app/src/migrations/02-create-search-indices.cypher
api_1    | Failed to run migration: /app/src/migrations/01-drop-search-indices.cypher
api_1    | Failed to run migration: /app/src/migrations/02-create-search-indices.cypher
api_1    | Error while running transactions: Failed to invoke procedure db.index.fulltext.drop: Caused by: java.lang.IllegalArgumentException: There is no such fulltext schema index: metadataSearchFields
api_1    | Successfully run 2 migrations
```

I guess the failure for 01 is because the index doesn't exist, however it also says "Failed to run migration 02", which should be the create index

**Comment by ChristiaanScheermeijer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/trompamusic/ce-api/pullRequest?prid=5244919&bid=17190401)

File: src/migrations.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import walkSync from 'walk-sync'
+import { readFileSync } from 'fs'
```
Comment (ChristiaanScheermeijer) [inline src/migrations.js]: Codacy Issue found: ['readFileSync' is defined but never used.](https://app.codacy.com/manual/trompamusic/ce-api/pullRequest?prid=5244919)

**Comment by ChristiaanScheermeijer** [comment]
@alastair I needed to change the migrations runner because it seems not possible to run schema queries conditionally. I've tried multiple queries like the one below to prevent this change but didn't succeed.

```cypher
CALL db.indexes() YIELD indexName WHERE indexName = 'metadataSearchFields' WITH count(*) as indexCount
CALL apoc.do.when(indexCount > 0, 'CALL db.index.fulltext.drop(\'metadataSearchFields\') RETURN true', 'RETURN false') YIELD value RETURN value
```

**Comment by alastair** [comment]
If I run a second migration file, I get the following error:

> api_1    | Failed to commit migrations: Cannot perform data updates in a transaction that has performed schema updates.

It looks like we'll have to start a new transaction for each file. I tried it and it works fine

**Comment by ChristiaanScheermeijer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/migrations.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/trompamusic/ce-api/pullRequest?prid=5244919&bid=17190401)

**Pull request merged**